arbor/backends/gpu/shared_state.cpp

```diff
@@ -86,6 +87,81 @@ void ion_state::reset() {
     memory::copy(init_eX_, eX_);
 }
 
+// istim_state methods:
+
+istim_state::istim_state(const fvm_stimulus_config& stim) {
+    using util::assign;
+
+    // Translate instance-to-CV index from stim to istim_state index vectors.
+    std::vector<fvm_index_type> accu_index_stage;
+    assign(accu_index_stage, util::index_into(stim.cv, stim.cv_unique));
+
+    std::size_t n = accu_index_stage.size();
+    std::vector<fvm_value_type> envl_a, envl_t;
+    std::vector<fvm_index_type> edivs;
+
+    frequency_ = make_const_view(stim.frequency);
+
+    arb_assert(n==frequency_.size());
+    arb_assert(n==stim.envelope_time.size());
+    arb_assert(n==stim.envelope_amplitude.size());
+
+    edivs.reserve(n+1);
+    edivs.push_back(0);
+
+    for (auto i: util::make_span(n)) {
+        arb_assert(stim.envelope_time[i].size()==stim.envelope_amplitude[i].size());
+        arb_assert(util::is_sorted(stim.envelope_time[i]));
+
+        util::append(envl_a, stim.envelope_amplitude[i]);
+        util::append(envl_t, stim.envelope_time[i]);
+        edivs.push_back(fvm_index_type(envl_t.size()));
+    }
+
+    accu_index_ = make_const_view(accu_index_stage);
+    accu_to_cv_ = make_const_view(stim.cv_unique);
+    accu_stim_ = array(accu_index_.size());
+    envl_amplitudes_ = make_const_view(envl_a);
+    envl_times_ = make_const_view(envl_t);
+    envl_divs_ = make_const_view(edivs);
+
+    // Initial indices into envelope match partition divisions; ignore last (index n) element.
+    envl_index_ = envl_divs_;
+
+    // Initialize ppack pointers.
+    ppack_.accu_index = accu_index_.data();
+    ppack_.accu_to_cv = accu_to_cv_.data();
+    ppack_.frequency = frequency_.data();
+    ppack_.envl_amplitudes = envl_amplitudes_.data();
+    ppack_.envl_times = envl_times_.data();
+    ppack_.envl_divs = envl_divs_.data();
+    ppack_.accu_stim = accu_stim_.data();
+    ppack_.envl_index = envl_index_.data();
+    ppack_.time = nullptr; // These fields must be set in add_current() before queing kernel.
```

Comment:
1. Typo: queuing?
2. Why?
3. 'These' implies more than one, which exactly?

Reply:
1. Definitely ytpo.
2. Because otherwise they will be null? For time, it's because it is being flipped around with time_to every timestep, but for the others, it's just a division of responsibility: the shared state code knows what it is doing with the shared vectors it is creating (especially `time`), so let's leave it to that code to do the requisite setting.
3. This line and the following, but I can make that more clear.